flow/include/flow/FlowThread.h

```diff
@@ -0,0 +1,294 @@
+
+	bool isReady() {
+		ThreadSpinLockHolder holder(mutex);
+		return !queue.empty() || error.isValid();
+	}
+
+	bool isError() {
+		ThreadSpinLockHolder holder(mutex);
+		return queue.empty() && error.isValid();
+	} // the *next* thing queued is an error
+
+	bool hasError() {
+		ThreadSpinLockHolder holder(mutex);
+		return error.isValid();
+	} // there is an error queued
+
+	uint32_t size() {
+		ThreadSpinLockHolder holder(mutex);
+		return queue.size();
+	}
+
+	virtual T pop() {
+		ThreadSpinLockHolder holder(mutex);
+		if (queue.empty()) {
+			if (error.isValid())
+				throw error;
+			throw internal_error();
+		}
+		auto copy = std::move(queue.front());
+		queue.pop();
+		return copy;
+	}
+
+	template <class U>
+	void send(U&& value) {
+		if (error.isValid())
+			return;
+
+		addPromiseRef();
+		onMainThreadVoid([this, value = std::forward<U>(value)]() {
+			mutex.enter();
+			if (this->next != this) {
+				auto n = this->next;
+				mutex.leave();
+				n->fire(value);
+			} else {
+				this->queue.emplace(value);
+				mutex.leave();
+			}
+			this->delPromiseRef();
+		});
+	}
+
+	void sendError(Error err) {
+		if (error.isValid())
+			return;
+
+		ASSERT(this->error.code() != error_code_success);
+		this->error = err;
+
+		addPromiseRef();
+		onMainThreadVoid([this, err]() {
+			// end_of_stream error is "expected", don't terminate reading stream early for this
+			SingleCallback<T>* n = nullptr;
+			{
+				ThreadSpinLockHolder holder(mutex);
+				if (this->shouldFireImmediatelyUnsafe()) {
+					n = this->next;
+				}
+			}
+
+			if (n)
+				n->error(err);
+
+			this->delPromiseRef();
+		});
+	}
+
+	void addPromiseRef() {
+		ThreadSpinLockHolder holder(mutex);
+		promises++;
+	}
+
+	void addFutureRef() {
+		ThreadSpinLockHolder holder(mutex);
+		futures++;
+	}
+
+	void delPromiseRef() {
+		mutex.enter();
+		if (!--promises) {
+			if (futures) {
+				mutex.leave();
+				sendError(broken_promise());
+			} else {
+				mutex.leave();
+				destroy();
+			}
+		} else {
+			mutex.leave();
+		}
+	}
+
+	void delFutureRef() {
+		mutex.enter();
+		if (!--futures) {
+			if (promises) {
+				mutex.leave();
+				cancel();
+			} else {
+				mutex.leave();
+				destroy();
+			}
+		} else {
+			mutex.leave();
+		}
+	}
+
+	int getFutureReferenceCount() {
+		ThreadSpinLockHolder holder(mutex);
+		return futures;
+	}
+
+	int getPromiseReferenceCount() {
+		ThreadSpinLockHolder holder(mutex);
+		return promises;
+	}
+
+	void addCallbackAndDelFutureRef(SingleCallback<T>* cb) {
+		ThreadSpinLockHolder holder(mutex);
+		ASSERT(SingleCallback<T>::next == this);
+		cb->insert(this);
+	}
+
+	virtual void destroy() { delete this; }
+	virtual void cancel() {}
+	virtual void unwait() override { delFutureRef(); }
+	virtual void fire(T const&) override { ASSERT(false); }
+	virtual void fire(T&&) override { ASSERT(false); }
+
+protected:
+	bool shouldFireImmediatelyUnsafe() { return SingleCallback<T>::next != this; }
+};
+
+template <class T>
+class ThreadFutureStream {
```

Comment:
Could this not subclass FutureStream?

Reply:
Not without making `ThreadNotifiedQueue` a subclass of `NotifiedQueue` and marking all `NotifiedQueue` methods to virtual, which would in turn change the original primitives (add vtable etc) which which we want to avoid. And that would still require changing Flow compiler (and more so than we do now) because its not using generating pointer to future but copying in stack. Deriving `ThreadNotifiedQueue` itself will not be super useful as anything non-trivial needs to re-done anyway, as you want to avoid things like destroy(), cancel(), sendError() while inside a lock. 

There are ways to use template and other meta-programming magic to avoid dynamic dispatch and code duplication, however not without changing that part of the code.

Reply:
Sounds good. Maybe add comment to this effect on this new class that looks very much like FutureStream explaining why there is no inheritance. Thanks.